### ISSUE TITLE
k8s.io/client-go: misuse of http.DefaultClient in rest.HTTPClientFor

### ISSUE THREAD

**Issue by achille-roussel** [kubernetes/kubernetes#119418]
### What happened?

Hello,

I am opening this issue to discuss a potential misuse of `http.DefaultClient` in `k8s.io/client-go/rest.HTTPClientFor`. The function appears to make the assumption that `http.DefaultClient` will always use `http.DefaultTransport`, which may not be the case.

For context, this is the function in question: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/transport.go#L29-L48

### What did you expect to happen?

I would expect to be able to change `http.DefaultTransport` without implicitly impacting the behavior of Kubernetes clients that have a different configuration.

### How can we reproduce it (as minimally and precisely as possible)?

Take this code as example:
```go
package main

func init() {
    // change the default client transport's to something custom
    http.DefaultClient.Transport = &http.Transport{
        ...
    }
}

func main() {
   ...
}
```

In this example, invocations of `rest.HTTPClientFor` which resolve the transport to `http.DefaultTransport` will mistakenly assume that they can use the default client because the configuration of `http.DefaultClient.Transport` may not match `http.DefaultTransport` anymore.

### Anything else we need to know?

The check to fallback to `http.DefaultClient` in `rest.HTTPClientFor` seems to be intended as an optimization. Fixing the implementation could look like this:
```go
if transport != http.DefaultClient.Transport || (transport == http.DefaultTransport && http.DefaultClient.Transport != nil) || config.Timeout > 0 {
    httpClient = &http.DefaultClient{
        Transport: transport,
        Timeout:   config.Timeout,
    }
} else {
    httpClient = http.DefaultClient
}
```

However, this fix example highlights the tight coupling between this optimization and the inner implementation of `http.Client`. I would like to suggest removing it and always creating a new `http.Client` regardless of which transport is resolved from the configuration to simplify the code and avoid regressions that could be difficult to anticipate or track down.

To summarize, the function would be modified to always do:
```go
return &http.Client{
    Transport: transport,
    Timeout:   config.Timeout,
}
```

Happy to submit the fix if there is a consensus on the issue.

### Kubernetes version

All recent k8s.io/client-go versions are affected

### Cloud provider

N/A

### OS version

All OS versions are affected

**Comment by achille-roussel** [comment]
Thanks for marking the issue as accepted!

@seans3 I noticed you got assigned to the issue, does it mean you are in charge of the implementation? I'm happy to send a PR for review if help is needed.

**Comment by aojea** [comment]
I'm not sure if is it even possible that ` transport == http.DefaultTransport` here

https://github.com/kubernetes/kubernetes/blob/99190634ab252604a4496882912ac328542d649d/staging/src/k8s.io/client-go/rest/transport.go#L33-L38C1
